### PR TITLE
Add backend for Hashicorp Vault (alternative approach)

### PR DESCRIPTION
**Pull request by candlerb**
## Proposed changes

This is a more generic solution to PR #668 - for consideration.

In this PR, the OpenIdConnectAuth class is enhanced so that it can be instantiated directly, as a "generic" OIDC connector.

To enable this, it also makes the OAuth2 authentication mechanism configurable (i.e. form post or HTTP Basic), and uses OIDC discovery to select this automatically.

Vault then becomes an empty subclass of OpenIdConnectAuth, since it's an out-of-the-box OIDC provider.

## Types of changes

Please check the type of change your PR introduces:

- [ ] Release (new release request)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code style update (PEP8, lint, formatting, renaming, etc)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build related changes (build process, tests runner, etc)
- [ ] Other (please describe):

I believe this *shouldn't* break existing code, and all tests pass, but it would be wise to check the other OpenIdConnectAuth subclasses just in case.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code._

- [X] Lint and unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works

## Other information

This PR also includes a fix for okta test which surfaced as part of this change. I believe this should only fix the test, not the actual behavior of the okta backend.

### CODE REVIEW

**Review by davidhalter — CHANGES_REQUESTED** [review]
Thanks for trying to contribute @candlerb 

I'm personally against this big list of changes that affect everything a lot of providers. I think we should go one step smaller and deal with individual problems as they arise. What is the issue for Vault here? What are you trying to solve?

The project is in maintenance mode and I feel like changes like this one are dangerous if there isn't a clear benefit to them. I won't merge for now and I would ask other maintainers not to merge as without discussing this first.

**Comment by candlerb** [comment]
Now I remember the other problem. In `OpenIdConnectAuth `, `OIDC_ENDPOINT` is a class variable, not a setting.

You can spoof it by turning it into a property (which is what I did in #668 to minimise other changes):

```
    @property
    def OIDC_ENDPOINT(self):
        return self.setting('OIDC_ENDPOINT')
```

This works, but it seems pretty ugly to me. It would probably be better just to have a fresh `GenericOpenIdConnectProvider` which doesn't inherit from `OpenIdConnectAuth` at all, even though that involves duplicating code.

The nearest existing backend I found with configurable endpoint URL is the [Keycloak backend](https://github.com/python-social-auth/social-core/blob/master/social_core/backends/keycloak.py), which gets around this problem because it doesn't subclass from `OpenIdConnectAuth`, it subclasses `BaseOAuth2`. However that means it loses all the auto-discovery: you can't just point at Keycloak's base url and pick up `.well-known/openid-configuration`. Instead you have to configure a load of settings by hand (`PUBLIC_KEY`, `AUTHORIZATION_URL`, `ACCESS_TOKEN_URL`)

**Comment by davidhalter** [comment]
I have personally had the same issue. I was able to work around that by using a function that generates those classes and writes them to `globals()`. It's really ugly, but in general it's still the right approach with the code we currently have. I originally also wanted to go with the approach you are taking, but ultimately it's probably the wrong approach for what we have here.

If you are just writing an OpenID Connect compatible software that only needs OIDC, I would probably recommend to not use this library, because it's not very well maintained. There are probably other libraries that only do OpenID Connect and are therefore more correct when it comes to the standard.

In general I'm just worried about touching the OIDC pieces here, because they are already not that well maintained and almost nobody around here understands OIDC on a good level. I certainly do not, because it's just hard to understand and I'm kind of maintaining that part :-/.

**Comment by candlerb** [comment]
> If you are just writing an OpenID Connect compatible software that only needs OIDC, I would probably recommend to not use this library, because it's not very well maintained. There are probably other libraries that only do OpenID Connect and are therefore more correct when it comes to the standard.

It's the other way round: I wanted to take an application which already has python-social-auth integration ([Netbox](https://github.com/netbox-community/netbox/releases/tag/v3.1.0)), and connect it to an OIDC provider.

Netbox does offer an alternative way of doing this, via [remote authentication](https://netbox.readthedocs.io/en/stable/configuration/remote-authentication/) with HTTP headers, so I can instead use an Apache reverse proxy with mod_auth_openidc. I just thought that in fixing this use case for Netbox, it would be nice to make it available to other consumers of python-social-auth.

**Comment by nijel** [comment]
I don't see anything breaking in this (though my OpenID Connect experience is very limited). I think having generic OpenID Connect provider is good for reasons described by @candlerb - we use python-social-auth at Weblate and users do not have the choice for the authentication library (I agree that there might be better choices out there these days), but they want to have the choice in the authentication service. Therefore, I'd go and merge this.

It would be great, if @candlerb could prepare matching docs pull request in https://github.com/python-social-auth/social-docs/

**Comment by davidhalter** [comment]
Since I feel very much responsible for OpenID Connect stuff, please do not merge (otherwise I might have to revert). I will try to review some things again next week. My feeling still is that this is a bad approach (and probably breaks backwards compatibility), but I do not have the time to explain right now.

**Comment by nijel** [comment]
I've just gone again through this patch and I don't see anything breaking compatibility there. I think approach in having generic OpenID Connect backend is desired. For example [IdentityServer](https://github.com/DuendeSoftware/IdentityServer) could be integrated this way also.

**Comment by davidhalter** [comment]
I don't see how this is generic. If I wanted a generic OpenID Connect backend, I would want multiple ones. This is pretty common actually. The current approach is already generic, you just have to instantiate a class.

I still have to review the code better, though.

**Comment by nijel** [comment]
Yes, it's limited to a single OpenID Connect provider, but this is really what I'm looking for. Our users typically have internal authentication service and want to use that. Being able to specify that only in configuration without implementing a subclass for every service would be great.

**Comment by davidhalter** [comment]
> Yes, it's limited to a single OpenID Connect provider, but this is really what I'm looking for. Our users typically have internal authentication service and want to use that.

I feel like the difference of inheriting from a class and setting a few attributes on it versus setting a few variables in a global namespace is minimal. Both of these ways can be documented, will be around the same line count and with inheritance it is even possible to use multiple OIDC backends if that is ever wanted.

-------------

@candlerb

I kind of understand your approach with managing variables better now. Is there a way you could provide me with a way where in the OIDC standard you were looking up the basic auth stuff? I feel like the big problem is that we currently do not support that and the OIDC standard is really hard to get right...

**Comment by candlerb** [comment]
> Is there a way you could provide me with a way where in the OIDC standard you were looking up the basic auth stuff? I feel like the big problem is that we currently do not support that and the OIDC standard is really hard to get right...

Sure. When you want to exchange a code for an ID token (at the Token endpoint), OIDC provides two ways of the client authenticating itself: via HTTP Basic Authentication, or via client_id and client_secret fields in the POST request body.

This is described in [section 3.1.3.2](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequestValidation) and [section 9](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) of the OIDC Core 1.0 specification.

This in turn refers to [section 2.3.1](https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1) of the OAuth 2.0 specification, RFC 6749. Here it is made explicit that the authorization server *must* implement HTTP Basic Auth, and *may* also implement the POST body fields (but the latter is *not recommended*).

By supporting only basic auth but not POST, Vault is compliant to the standard. However, prior to this patch, PSA's oauth.py (and hence also open_id_connect.py) implements only the POST option.

There are some subclasses of BaseOAuth2 which override some methods to do basic auth - see for example RedditOAuth2.

File: social_core/backends/open_id_connect.py
Comment (davidhalter) [inline social_core/backends/open_id_connect.py:56]: Shouldn't this be by default `'client_secret_post'`?

**Review by davidhalter — COMMENTED** [review]
Sorry for the delay again. I'm pretty busy and this is not easy to understand.

I feel like I understand the OpenID Connect part now, thanks for linking the relevant documents.

I feel like since @nijel wants the settings changes, this is also fine (although I personally prefer the subclassing stuff). Is there documentation needed @nijel ?

File: social_core/backends/open_id_connect.py
Comment (candlerb) [reply social_core/backends/open_id_connect.py:56]: Agreed. The logic says "anything other than client_secret_basic is client_secret_post" (hence the boolean function `use_basic_auth` in oauth.py), but it doesn't harm to document it properly.

**Comment by nijel** [comment]
Yes, documentation for new backends is definitely needed in https://github.com/python-social-auth/social-docs/

**Comment by candlerb** [comment]
I made some docs a while ago here: https://github.com/python-social-auth/social-docs/compare/master...candlerb:candlerb/oidc

I didn't make a PR as it was unclear whether these oidc changes were accepted, either in current form or differently.

File: social_core/backends/open_id_connect.py
Comment (davidhalter) [reply social_core/backends/open_id_connect.py:56]: Ok, I guess I agree. Can you maybe just copy paste your own comment `unspecified, let the server choose via openid autoconfiguration` and add that as a comment? I guess in that case it's clear why it's empty.

Comment (candlerb) [reply social_core/backends/open_id_connect.py:56]: I guess so, although the same also applies to all these parameters:

```
    ID_TOKEN_ISSUER = ''
    ACCESS_TOKEN_URL = ''
    AUTHORIZATION_URL = ''
    REVOKE_TOKEN_URL = ''
    USERINFO_URL = ''
    JWKS_URI = ''
```

Comment (candlerb) [reply social_core/backends/open_id_connect.py:57]: I have added a comment covering all these options

Comment (davidhalter) [reply social_core/backends/open_id_connect.py:57]: Awesome, thanks!

**Review by davidhalter — APPROVED** [review]

**Comment by davidhalter** [comment]
To me this looks good now, thanks a lot. I think @nijel should approve all the documentation stuff, I have no idea about that.

**Comment by nijel** [comment]
I've made the PR for the docs here: https://github.com/python-social-auth/social-docs/pull/109

**Comment by nijel** [comment]
Merged, thanks for your contribution!

**Pull request merged**
